### PR TITLE
AC-575 : Reset Password

### PR DESCRIPTION
**Pull request by VibhorChinda**
## Description of what I changed

I have created a dialog box where the user can add his/her email id 
I have created a method for the user to change his/her password using retrofit

## Issue I worked on

JIRA Issue: https://issues.openmrs.org/browse/AC-575

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [X] All new and existing **tests passed**.

- [X] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by VibhorChinda** [comment]
hello

**Comment by VibhorChinda** [comment]
@f4ww4z please have a look at the PR.
I have created a retrofit call to the API as mentioned in https://wiki.openmrs.org/pages/viewpage.action?pageId=216367317

But for some case i am getting error code 500 i.e internal server error is the API working fine ?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java [hunk 1]
```diff
@@ -288,6 +302,12 @@ public void showLoadingAnimation() {
         mSpinner.setVisibility(View.VISIBLE);
     }
 
+    @Override
+    public void showToastMessage(String message)
+    {
+        Toast.makeText(mOpenMRS, message, Toast.LENGTH_SHORT).show();
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java]: Did you see [ToastUtil](https://github.com/openmrs/openmrs-contrib-android-client/blob/master/openmrs-client/src/main/java/org/openmrs/mobile/utilities/ToastUtil.java) ?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java [hunk 2]
```diff
@@ -262,14 +265,25 @@ private void setSyncButtonState(boolean syncEnabled) {
     }
 
     public void forgotPassword() {
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java]: Add this to the `LoginContract`. Rename to `showForgotPasswordDialog`

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java [hunk 3]
```diff
@@ -262,14 +265,25 @@ private void setSyncButtonState(boolean syncEnabled) {
...
-        CustomDialogBundle bundle = new CustomDialogBundle();
-        bundle.setTitleViewMessage(getString(R.string.forgot_dialog_title));
-        bundle.setTextViewMessage(getString(R.string.forgot_dialog_message));
-        bundle.setLeftButtonAction(CustomFragmentDialog.OnClickAction.DISMISS);
-        bundle.setLeftButtonText(getString(R.string.forgot_button_ok));
-        ((LoginActivity) this.getActivity()).createAndShowDialog(bundle, ApplicationConstants.DialogTAG.LOGOUT_DIALOG_TAG);
+        AlertDialog.Builder alertDialog = new AlertDialog.Builder(getContext());
+        View view = getLayoutInflater().inflate(R.layout.alert_dialog_text_input_layout,null);
+        TextInputEditText emailTextField = view.findViewById(R.id.forgot_password_dialog_text_field);
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java]: Move line [269 & 270](https://github.com/openmrs/openmrs-contrib-android-client/pull/654/files#diff-c13748c0f5916618a7679770cb28b04aR269-R270) up 1, so the `alertDialog` can use the builder pattern. e.g.
```java
new AlertDialog.Builder(...)
   .setTitle(...)
   .setView(...)
   .setPositiveButton(...)
```

Rename `alertDialog` according to what the dialog shows.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java [hunk 4]
```diff
@@ -262,14 +265,25 @@ private void setSyncButtonState(boolean syncEnabled) {
...
+        alertDialog.setView(view);
+        alertDialog.setPositiveButton(getString(R.string.forgot_password_dialog_positive_button), (dialog, which) -> {
+            if(emailTextField.getText() != null && emailTextField.getText().length() > 0) {
+                mPresenter.resetPassword(emailTextField.getText().toString());
+            }
+            else
+            {
+                showToastMessage(getString(R.string.email_text_field_empty_message));
+            }
+        }).setNegativeButton(R.string.dialog_button_cancel,((dialog, which) -> {
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java]: Why not `cancel` the dialog?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java [hunk 5]
```diff
@@ -262,14 +265,25 @@ private void setSyncButtonState(boolean syncEnabled) {
     }
 
     public void forgotPassword() {
-        CustomDialogBundle bundle = new CustomDialogBundle();
-        bundle.setTitleViewMessage(getString(R.string.forgot_dialog_title));
-        bundle.setTextViewMessage(getString(R.string.forgot_dialog_message));
-        bundle.setLeftButtonAction(CustomFragmentDialog.OnClickAction.DISMISS);
-        bundle.setLeftButtonText(getString(R.string.forgot_button_ok));
-        ((LoginActivity) this.getActivity()).createAndShowDialog(bundle, ApplicationConstants.DialogTAG.LOGOUT_DIALOG_TAG);
+        AlertDialog.Builder alertDialog = new AlertDialog.Builder(getContext());
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java]: This should be an  `AlertDialog` object, with `create()` called after building it.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java [hunk 6]
```diff
@@ -89,6 +91,24 @@ public void subscribe() {
         // This method is intentionally empty
     }
 
+    public void resetPassword(final String email)
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: This method should override the one you specified in `LoginContract`

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java [hunk 7]
```diff
@@ -89,6 +91,24 @@ public void subscribe() {
...
+    {
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: Did you use our code style? This should be in the same line as the method name. Line 94

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java [hunk 8]
```diff
@@ -89,6 +91,24 @@ public void subscribe() {
...
+        restApi.resetPassword(email).enqueue(new Callback<Void>() {
+            @Override
+            public void onResponse(Call<Void> call, Response<Void> response) {
+                if (response.isSuccessful())
+                    loginView.showToastMessage("Please check your inbox.");
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: Put this in `strings.xml` . Mention the email address too.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java [hunk 9]
```diff
@@ -89,6 +91,24 @@ public void subscribe() {
...
+                else
+                    loginView.showToastMessage("Faced some error please try again.");
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: Put in `strings.xml`

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java [hunk 10]
```diff
@@ -89,6 +91,24 @@ public void subscribe() {
...
+            }
+
+            @Override
+            public void onFailure(Call<Void> call, Throwable t) {
+                loginView.showToastMessage("Faced some error please try later.");
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: Put in `strings.xml` , show the user exactly what error. Use the parameter `call` and `t`.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java (hunk 9 above)
Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginPresenter.java]: Did you see our [Java Conventions](https://wiki.openmrs.org/display/docs/Java+Conventions) regarding if statements?

File: openmrs-client/src/main/res/layout/alert_dialog_text_input_layout.xml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    android:padding="@dimen/padding_16">
+
+    <com.google.android.material.textfield.TextInputLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:focusableInTouchMode="true"
+        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox">
+
+        <com.google.android.material.textfield.TextInputEditText
+            android:id="@+id/forgot_password_dialog_text_field"
```
Comment (f4ww4z) [inline openmrs-client/src/main/res/layout/alert_dialog_text_input_layout.xml]: See our [layout files](https://github.com/openmrs/openmrs-contrib-android-client/tree/master/openmrs-client/src/main/res/layout), and see the conventions for `id` names.

File: openmrs-client/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -32,11 +32,15 @@
     <string name="req_star"><![CDATA[<font color=\'#cc0029\'> * </font>]]></string>
     <string name="warning_dialog_title">Warning</string>
     <string name="warning_lost_data_dialog">Changing the URL, username or password will erase all the data. Are you sure?</string>
-    <string name="forgot_dialog_title">Unable to login?</string>
+    <string name="forgot_dialog_title">Forgot Password ?</string>
```
Comment (f4ww4z) [inline openmrs-client/src/main/res/values/strings.xml]:
```suggestion
    <string name="forgot_dialog_title">Forgot Password?</string>
```

File: openmrs-client/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -32,11 +32,15 @@
...
     <string name="forgot_dialog_message">Please contact the system administrator</string>
     <string name="forgot_button_ok">Dismiss</string>
     <string name="login_offline">Offline</string>
     <string name="login_online">Online</string>
+    <string name="enter_registered_email">Enter email</string>
+    <string name="forgot_password_dialog_positive_button">DONE</string>
+    <string name="email_text_field_empty_message">Please fill the email text field.</string>
```
Comment (f4ww4z) [inline openmrs-client/src/main/res/values/strings.xml]:
```suggestion
    <string name="email_text_field_empty_message">Please enter your email.</string>
```

File: openmrs-client/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -274,7 +278,7 @@
     <string name="birthdate_hint">birthdate:</string>
     <string name="address_hint">address:</string>
     <string name="demographics">Demographics</string>
-    <string name="unable_to_login">Unable to login?</string>
+    <string name="unable_to_login">Forgot Password?</string>
```
Comment (f4ww4z) [inline openmrs-client/src/main/res/values/strings.xml]: Rename the string also.

Comment (VibhorChinda) [reply openmrs-client/src/main/res/values/strings.xml]: Ok @f4ww4z sir

File: openmrs-client/src/main/res/layout/alert_dialog_text_input_layout.xml (hunk 11 above)
Comment (VibhorChinda) [reply openmrs-client/src/main/res/layout/alert_dialog_text_input_layout.xml]: Ok @f4ww4z sir will keep this convention in mind later also.

**Comment by VibhorChinda** [comment]
@f4ww4z sir please check the PR.

**Comment by NASA12345** [comment]
Hello, I completely agree with sir @VibhorChinda. Sir, @f4ww4z please check 
Pull request again to improve. I am also getting the error 500 as the internal server error. I suggest you to improve it to make it better

**Comment by Devanshk9** [comment]
@VibhorChinda sir please merge 2 commits into one for proceeding futher

**Comment by VibhorChinda** [comment]
> @VibhorChinda sir please merge 2 commits into one for proceeding futher

will surely squash all my commits once my PR has no more changes to be done.

**Comment by f4ww4z** [comment]
@VibhorChinda  Did you see this https://wiki.openmrs.org/pages/viewpage.action?pageId=216367317 wiki? It's got all the steps needed to setup the server. Please run it on the demo server, if possible, and show your progress here.

**Comment by f4ww4z** [comment]
@VibhorChinda what is the progress of this? Were you able to setup a local server and implement the password reset feature in the reference application?

**Comment by VibhorChinda** [comment]
> @VibhorChinda what is the progress of this? Were you able to setup a local server and implement the password reset feature in the reference application?

@f4ww4z sir, i was able to setup the server locally but the api seems to be not working.
This is discussion link where i described what all progress i made and what error i was facing. 
https://talk.openmrs.org/t/testing-the-reset-password-feature-for-android-client/26798

**Comment by f4ww4z** [comment]
> @f4ww4z sir, i was able to setup the server locally but the api seems to be not working.
> This is discussion link where i described what all progress i made and what error i was facing.
> https://talk.openmrs.org/t/testing-the-reset-password-feature-for-android-client/26798

@VibhorChinda I will try setting up a local server with the password reset feature too and let you know the result.

**Comment by f4ww4z** [comment]
> @VibhorChinda I will try setting up a local server with the password reset feature too and let you know the result.

@VibhorChinda Since I didn't get the time to set up a local server with this feature, and because of the inactivity, I'm closing this issue *for now*. Ping me back if you want to resume working on it.

**Pull request closed without merging**